File: src/main/java/com/github/bjoernpetersen/jmusicbot/playback/Player.java
package com.github.bjoernpetersen.jmusicbot.playback;

import com.github.bjoernpetersen.jmusicbot.Loggable;
import com.github.bjoernpetersen.jmusicbot.Song;
import com.github.bjoernpetersen.jmusicbot.playback.PlayerState.State;
import com.github.bjoernpetersen.jmusicbot.provider.Suggester;
import com.google.common.util.concurrent.ThreadFactoryBuilder;
import java.io.Closeable;
import java.io.IOException;
import java.util.HashSet;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;
import java.util.logging.Logger;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

public final class Player implements Loggable, Closeable {

  @Nonnull
  private final Logger logger;

  @Nonnull
  private final ExecutorService executorService;

  @Nonnull
  private final Consumer<SongEntry> songPlayedNotifier;

  @Nonnull
  private final Queue queue;
  @Nullable
  private final Suggester suggester;

  @Nonnull
  private final Lock stateLock;
  @Nonnull
  private PlayerState state;
  @Nonnull
  private Playback playback;

  @Nonnull
  private final Set<PlayerStateListener> stateListeners;

  public Player(@Nonnull Consumer<SongEntry> songPlayedNotifier, @Nullable Suggester suggester) {
    this.logger = createLogger();

    this.executorService = Executors.newSingleThreadExecutor(
        new ThreadFactoryBuilder()
            .setDaemon(true)
            .setNameFormat("playerPool-%d")
            .build()
    );

    this.songPlayedNotifier = songPlayedNotifier;

    this.queue = new Queue();
    this.suggester = suggester;

    this.stateLock = new ReentrantLock();
    this.state = PlayerState.stop();
    this.playback = DummyPlayback.INSTANCE;

    this.stateListeners = new HashSet<>();

    executorService.submit(this::autoPlay);
  }

  @Override
  @Nonnull
  public Logger getLogger() {
    return logger;
  }

  public void addListener(@Nullable PlayerStateListener listener) {
    if (listener != null) {
      stateListeners.add(listener);
    }
  }

  public void removeListener(@Nullable PlayerStateListener listener) {
    if (listener != null) {
      stateListeners.remove(listener);
    }
  }

  @Nonnull
  public Queue getQueue() {
    return queue;
  }

  @Nonnull
  public PlayerState getState() {
    return state;
  }

  private void setState(PlayerState state) {
    this.state = state;
    for (PlayerStateListener listener : stateListeners) {
      listener.onChanged(state);
    }
  }

  public void pause() {
    Lock stateLock = this.stateLock;
    stateLock.lock();
    try {
      logFinest("Pausing...");
      PlayerState state = getState();
      if (state.getState() == State.PAUSE) {
        logFinest("Already paused.");
        return;
      } else if (state.getState() != State.PLAY) {
        logFiner("Tried to pause player in state %s", state.getState());
        return;
      }
      playback.pause();
      setState(PlayerState.pause(state.getEntry().orElseThrow(IllegalStateException::new)));
    } finally {
      stateLock.unlock();
    }
  }

  public void play() {
    Lock stateLock = this.stateLock;
    stateLock.lock();
    try {
      logFinest("Playing...");
      PlayerState state = getState();
      if (state.getState() == State.PLAY) {
        logFinest("Already playing.");
        return;
      } else if (state.getState() != State.PAUSE) {
        logFiner("Tried to play in state %s", state.getState());
        return;
      }
      playback.play();
      setState(PlayerState.play(state.getEntry().orElseThrow(IllegalStateException::new)));
    } finally {
      stateLock.unlock();
    }
  }

  public void next() throws InterruptedException {
    PlayerState state = getState();
    Lock stateLock = this.stateLock;
    stateLock.lock();
    try {
      logFiner("Next...");
      PlayerState newState = getState();
      if (isSignificantlyDifferent(state, newState)) {
        logFinest("Skipping next call due to state change while waiting for lock.");
        return;
      }

      try {
        playback.close();
      } catch (Exception e) {
        logWarning(e, "Error closing playback");
      }

      Optional<QueueEntry> nextOptional = queue.pop();
      if (!nextOptional.isPresent() && suggester == null) {
        logFinest("Queue is empty. Stopping.");
        playback = DummyPlayback.INSTANCE;
        setState(PlayerState.stop());
        return;
      }

      SongEntry nextEntry;
      if (nextOptional.isPresent()) {
        nextEntry = nextOptional.get();
      } else {
        nextEntry = new SuggestedSongEntry(suggester.suggestNext());
      }

      Song nextSong = nextEntry.getSong();
      songPlayedNotifier.accept(nextEntry);
      logFiner("Next song is: " + nextSong);
      try {
        playback = nextSong.getPlayback();
      } catch (IOException e) {
        logWarning(e, "Error creating playback");

        setState(PlayerState.error());
        playback = DummyPlayback.INSTANCE;
        return;
      }

      playback.setPlaybackStateListener(playbackState -> {
        logFinest("Playback state update: %s", playbackState);
        switch (playbackState) {
          case PLAY:
            play();
            break;
          case PAUSE:
            pause();
            break;
          default:
            throw new IllegalArgumentException("Unknown PlaybackState: " + playbackState);
        }
      });

      setState(PlayerState.pause(nextEntry));
      play();
    } finally {
      stateLock.unlock();
    }
  }

  private boolean isSignificantlyDifferent(@Nonnull PlayerState state, @Nonnull PlayerState other) {
    Optional<SongEntry> stateSong = state.getEntry();
    Optional<SongEntry> otherSong = other.getEntry();
    // TODO check for correctness
    return (otherSong.isPresent() != stateSong.isPresent())
        || otherSong.isPresent() && !state.getEntry().equals(other.getEntry());
  }

  private void autoPlay() {
    try {
      PlayerState state = getState();
      logFinest("Waiting for song to finish");
      playback.waitForFinish();
      logFinest("Waiting done");

      Lock stateLock = this.stateLock;
      stateLock.lock();
      try {
        // Prevent auto next calls if next was manually called
        if (isSignificantlyDifferent(getState(), state)) {
          logFinest("Skipping auto call to next()");
        } else {
          logFinest("Auto call to next()");
          next();
        }
      } finally {
        stateLock.unlock();
      }

      executorService.submit(this::autoPlay);
    } catch (InterruptedException e) {
      logFine("autoPlay interrupted", e);
    }
  }

  @Override
  public void close() throws IOException {
    try {
      executorService.shutdownNow();
      playback.close();
    } catch (Exception e) {
      throw new IOException(e);
    }
  }
}

File: src/main/java/com/github/bjoernpetersen/jmusicbot/playback/Player.kt
package com.github.bjoernpetersen.jmusicbot.playback

import com.github.bjoernpetersen.jmusicbot.Song
import com.github.bjoernpetersen.jmusicbot.playback.PlaybackStateListener.PlaybackState
import com.github.bjoernpetersen.jmusicbot.provider.BrokenSuggesterException
import com.github.bjoernpetersen.jmusicbot.provider.Suggester
import com.google.common.util.concurrent.ThreadFactoryBuilder
import mu.KotlinLogging
import java.io.Closeable
import java.io.IOException
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock
import java.util.function.Consumer
import kotlin.concurrent.withLock

class Player(private val songPlayedNotifier: Consumer<SongEntry>, private val suggester: Suggester?) :
    Closeable {

  private val logger = KotlinLogging.logger {}
  private val autoPlayer: ExecutorService = Executors.newSingleThreadExecutor(
      ThreadFactoryBuilder()
          .setDaemon(true)
          .setNameFormat("playerPool-%d")
          .build()
  )

  /**
   * This player's queue.
   */
  val queue: Queue = Queue()

  private val stateLock: Lock = ReentrantLock()
  /**
   * The current state of this player. This might be play, pause, stop or error.
   */
  var state: PlayerState = StopState()
    private set(value) {
      field = value
      for (listener in stateListeners) {
        listener.onChanged(value)
      }
    }
  private var playback: Playback = DummyPlayback

  private val stateListeners: MutableSet<PlayerStateListener> = HashSet()

  init {
    autoPlayer.submit { this.autoPlay() }
    queue.addListener(object : QueueChangeListener {
      override fun onAdd(entry: QueueEntry) {
        entry.song.load()
      }

      override fun onRemove(entry: QueueEntry) {}

      override fun onMove(entry: QueueEntry, fromIndex: Int, toIndex: Int) {}
    })

    if (suggester != null) {
      addListener(PlayerStateListener { _ -> preloadSuggestion(suggester) })
    }
  }

  private fun preloadSuggestion(suggester: Suggester) {
    if (queue.isEmpty) {
      val suggestions: List<Song>
      try {
        suggestions = suggester.getNextSuggestions(1)
      } catch (e: BrokenSuggesterException) {
        return
      }

      suggestions[0].load()
    }
  }

  /**
   * Adds a [PlayerStateListener] which will be called everytime the [state] changes.
   */
  fun addListener(listener: PlayerStateListener) {
    stateListeners.add(listener)
  }

  /**
   * Removes a [PlayerStateListener] previously registered with [addListener].
   */
  fun removeListener(listener: PlayerStateListener) {
    stateListeners.remove(listener)
  }

  /**
   * Pauses the player.
   *
   * If the player is not currently playing anything, nothing will be done.
   *
   * This method blocks until the playback is paused.
   */
  @Throws(InterruptedException::class)
  fun pause() {
    stateLock.withLock {
      logger.debug("Pausing...")
      val state = state
      if (state is PauseState) {
        logger.trace("Already paused.")
        return
      } else if (state !is PlayState) {
        logger.info { "Tried to pause player in state $state" }
        return
      }
      playback.pause()
      this.state = state.pause()
    }
  }

  /**
   * Resumes the playback.
   *
   * If the player is not currently pausing, nothing will be done.
   *
   * This method blocks until the playback is resumed.
   */
  @Throws(InterruptedException::class)
  fun play() {
    stateLock.withLock {
      logger.debug("Playing...")
      val state = state
      if (state is PlayState) {
        logger.trace("Already playing.")
        return
      } else if (state !is PauseState) {
        logger.info { "Tried to play in state $state" }
        return
      }
      playback.play()
      this.state = state.play()
    }
  }

  /**
   * Plays the next song.
   *
   * This method will play the next song from the queue.
   * If the queue is empty, the next suggested song from the primary [suggester] will be used.
   * If there is no primary suggester, the player will transition into the [StopState].
   *
   * This method blocks until either a new song is playing or the StopState is reached.
   */
  @Throws(InterruptedException::class)
  fun next() {
    val state = this.state
    stateLock.withLock {
      logger.debug("Next...")
      val newState = this.state
      if (isSignificantlyDifferent(state, newState)) {
        logger.debug("Skipping next call due to state change while waiting for lock.")
        return
      }

      try {
        playback.close()
      } catch (e: Exception) {
        logger.warn(e) { "Error closing playback" }
      }

      val nextOptional = queue.pop()
      if (!nextOptional.isPresent && suggester == null) {
        logger.info("Queue is empty. Stopping.")
        playback = DummyPlayback
        this.state = StopState()
        return
      }

      val nextEntry: SongEntry = if (nextOptional.isPresent) {
        nextOptional.get()
      } else try {
        SuggestedSongEntry(suggester!!.suggestNext())
      } catch (e: BrokenSuggesterException) {
        logger.warn("Default suggester could not suggest anything. Stopping.")
        playback = DummyPlayback
        this.state = StopState()
        return
      }

      val nextSong = nextEntry.song
      songPlayedNotifier.accept(nextEntry)
      logger.debug("Next song is: $nextSong")
      try {
        playback = nextSong.playback
      } catch (e: IOException) {
        logger.warn(e) { "Error creating playback" }

        this.state = ErrorState()
        playback = DummyPlayback
        return
      }

      playback.setPlaybackStateListener { this.onPlaybackFeedback(it) }

      this.state = PauseState(nextEntry)
      play()
    }
  }

  private fun onPlaybackFeedback(feedback: PlaybackState) {
    logger.debug { "Playback state update: $feedback" }
    stateLock.withLock {
      val state = state
      when (feedback) {
        PlaybackStateListener.PlaybackState.PLAY -> if (state is PauseState) {
          this.state = state.play()
        }
        PlaybackStateListener.PlaybackState.PAUSE -> if (state is PlayState) {
          this.state = state.pause()
        }
      }
    }
  }

  private fun isSignificantlyDifferent(state: PlayerState, other: PlayerState): Boolean {
    val stateSong = state.entry
    val otherSong = other.entry
    // TODO check for correctness
    return ((otherSong == null) != (stateSong == null)) || ((otherSong != null) && (stateSong != otherSong))
  }

  private fun autoPlay() {
    try {
      val state = this.state
      logger.debug("Waiting for song to finish")
      playback.waitForFinish()
      logger.trace("Waiting done")

      stateLock.withLock {
        // Prevent auto next calls if next was manually called
        if (isSignificantlyDifferent(this.state, state)) {
          logger.debug("Skipping auto call to next()")
        } else {
          logger.debug("Auto call to next()")
          next()
        }
      }

      autoPlayer.submit { this.autoPlay() }
    } catch (e: InterruptedException) {
      logger.info("autoPlay interrupted", e)
    }
  }

  @Throws(IOException::class)
  override fun close() {
    try {
      autoPlayer.shutdownNow()
      playback.close()
    } catch (e: Exception) {
      throw IOException(e)
    }
  }
}

private object DummyPlayback : Playback {
  override fun play() {}

  override fun pause() {}

  @Throws(InterruptedException::class)
  override fun waitForFinish() {
    Thread.sleep(2000)
  }

  override fun close() {}
}
